### ISSUE TITLE
SqlTranslate: Handle trailing semi-colons in SQL statements in Oracle

### ISSUE THREAD

**Issue by anthonysena** [OHDSI/SqlRender#151]
Per discussion here: https://github.com/OHDSI/WebAPI/pull/725#discussion_r251248996, we're exploring a way to modify SqlTranslate to handle the case described in that issue.

**Comment by schuemie** [comment]
In DatabaseConnector I just [always call `splitSql` before sending a SQL statement](https://github.com/OHDSI/DatabaseConnector/blob/master/R/Sql.R#L335). I'm not sure it makes sense to for example add a `splitSql` call  to the end of `translateSql`, since the result may contain multiple SQL statements. 

Also note that Oracle doesn't allow trailing semicolons when sending SQL via JDBC, but requires trailing semicolons when for example using Oracle's native SQL client. So as far as Oracle has any logic, that logic suggests the removing of the trailing semicolons is part of the call to the DB, not the SQL statement itself.

In other words: what exactly are you proposing we do?

**Comment by chrisknoll** [comment]
> In other words: what exactly are you proposing we do?

Hey, Martijn. I'm not sure what the 'best' solution is.  Our challenge is that we don't want to run through all our code and force the splitSql before each call.  It's just a bit of weeding for us.  In addition, we're not sure that it would work: if you always split a sql and you might get mutliple statements back, then we can't just do queryBatch() (in WebAPI we're using JDBC directly) with multiple statements because you're expected to get a result set back from your 'presuably' simple single querySql.  But executingBatch assumes you're doing a batch of update/insert/delete statements not a single query because the result of that call is an array of integers representing number of rows updated from the batch (and it doesn't return a result set).

So, we're just looking for simplest non-breaking solution that we can depend on SqlRender to give us back a good result instead of coding it around in each application that uses SqlRender.  I don't want to over-complicate SqlRender any more than necessary either, but in the case of Oracle (where there's confusion about where a `;` separates a block of code vs. separating a series of SQL is making it hard.

**Comment by schuemie** [comment]
So SQL to translate could have 1 statement or more than 1 statement.

If there's 1 statement, then you could apply splitSql to remove the trailing semicolon.

If there's more than 1, you can't, not if the output is another single SQL string, because you could never split the individual statements after removing the semicolons.

I think what you're saying is that you'd like a convenience function for the cases where you are absolutely certain there's only 1 statement in the SQl that does remove the trailing semicolon.  I could add a function called something like `translateSingleStatementSql` that would essentially call `translateSql` and `splitSql` in order. The nice thing is that it could also throw an error if more than 1 statement is encountered in the provided SQL.

**Comment by chrisknoll** [comment]
That is an excellent idea, and I think the only work on our side would be to find out where we might be calling a query that we expect to have a single statement.

@pavgra : I see this specifically for where we use PreparedStatementRenderer.  A prepared statement is by definition a single sql expression that should be invoked (I don't believe it can be multiple statements, right?).  If we updated this area of the code to use a new method like `translateSingleStatement` that would handle all the semicolon issues with oracle.

**Comment by chrisknoll** [comment]
@schuemie , one point on all this:  the point of this is that in the case of oracle, statements do not end with ; when they are translated.  So, hopefully that issue didn't get lost in all this discussion: the problem is that in the case of single statements on oracle, when the mssql form has the trailing ;, the oracle form can't execute it with the ;.

**Comment by schuemie** [comment]
@chrisknoll: that was indeed what I’m implying. Splitting and removing the trailing semicolon (for Oracle) are intertwined in the splitSql function. (Could separate splitting and removing semicolons into two functions, but I see no viable use case for that). The goal of the translateSingleStatementSql function is to remove the trailing semicolon for Oracle (and throw an error if more than 1 statement is found).

**Comment by chrisknoll** [comment]
I'm, I'm good with that, @pavgra , are you good with that?

**Comment by pavgra** [comment]
Ok, let's try such approach

**Comment by schuemie** [comment]
Ok, I'll try and implement it in develop, and you guys can take a look

**Comment by schuemie** [comment]
Let me know if those changes are ok, and I'll merge into master.

**Comment by chrisknoll** [comment]
Will do, thank you.
